Automatically exclude missing channels

alurban opened the pull request:
This PR adds a new utility under `gwdetchar.io.datafind` called `remove_missing_channels`. Whenever a list of channels is requested to `get_data`, this tool is used to exclude those channels that are missing from the first or last frame file within the given cache, throwing a loud warning showing exactly which channels are being ignored.

Two things to note:
* If NDS is used instead, `remove_missing_channels` is not invoked, meaning the call will still fail if channels cannot be found over the server.
* If only a single channel is passed (not a list), but that channel cannot be found in requested local frame files, then the call will still fail.

**Upshot:** Since all tools use `get_data` to retrieve timeseries data, this utility will cut down on the number of jobs that fail because channels are requested that cannot be found for the given time range, _except_ when using NDS or reading one channel at a time. In the latter case, users can easily work around this by wrapping the function call in a `try-except` block.

This fixes #150.

cc @duncanmmacleod, @areeda, @tjma12, @macedo22, @uberpye, @jrsmith02

duncanmmacleod commented on gwdetchar/io/datafind.py:
```diff
@@ -70,6 +75,47 @@ def check_flag(flag, gpstime, duration, pad):
+    # work out which channels to keep, and which to reject
+    keep = list(set(channels) & available)
+    reject = list(set(channels) - set(keep))
+    if reject:
```
This clause can be removed.

alurban replied:
good point, I don't know why that didn't occur to me....

duncanmmacleod commented on gwdetchar/io/datafind.py:
```diff
@@ -70,6 +75,47 @@ def check_flag(flag, gpstime, duration, pad):
+        available.update(io_gwf.iter_channel_names(gwfcache[-1]))
+    # work out which channels to keep, and which to reject
+    keep = list(set(channels) & available)
+    reject = list(set(channels) - set(keep))
```
It would be more efficient to do something like this:

```python
    channels = set(channels)
    keep = channels & available
    reject = channels - keep
```

and then just `return list(keep)` at the end.

alurban replied:
Yep, done in the latest commit

duncanmmacleod commented on bin/gwdetchar-overflow:
```diff
@@ -200,10 +201,14 @@ for dcuid in args.dcuid:
+                    )
+                except Exception:
+                    traceback.print_exc()
+                    gprint("    Problem encountered, skipping {}".format(ch))
```
Can you just use `warnings` here instead:

```python
            except Exception as exc:
                warnings.warn("skipping {}: [{}] {}".format(ch, type(exc), str(exc))
```

alurban replied:
Sure, this is done in the latest commit

alurban commented:
@duncanmmacleod, I've addressed your review comments (so far), and resolved a few build test failures

alurban commented:
@duncanmmacleod, for completeness, [**here**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/hveto/day/20190108/scans/1230955301.88/) (requires `LIGO.ORG`) is a completed omega scan of a glitch on 8 January that would have failed without these changes because several channels have disappeared since then. [**Here**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/hveto/day/20190108/scans/logs/gwdetchar-omega_1230955301_88.error) is the stderr output from that scan, showing the various warning statements.

duncanmmacleod commented on bin/gwdetchar-omega:
```diff
@@ -223,9 +221,9 @@ for block in blocks.values():
-        except Exception:
-            traceback.print_exc()
-            print('    Problem encountered, skipping {}'.format(channel.name))
+        except Exception as exc:
```
Please identify the expected error modes here, rather than just `Exception`, since we only want to catch things associated with channels missing from the source data.

duncanmmacleod commented on bin/gwdetchar-overflow:
```diff
@@ -200,10 +201,14 @@ for dcuid in args.dcuid:
+                        data[ch],
+                        cumulative=True,
+                    )
+                except Exception as exc:
```
Please identify the expected error modes here, rather than just `Exception`, since we only want to catch things associated with channels missing from the source data.

alurban commented on bin/gwdetchar-omega:
```diff
@@ -223,9 +221,9 @@ for block in blocks.values():
-            print('    Problem encountered, skipping {}'.format(channel.name))
+        except Exception as exc:
+            warnings.warn("Skipping {}: [{}] {}".format(
+                channel, type(exc), str(exc)), UserWarning)
```
Good catch, this is done in the last commit.

alurban commented on gwdetchar/io/datafind.py:
```diff
@@ -70,6 +73,53 @@ def check_flag(flag, gpstime, duration, pad):
+    # get available channels from the first and last frame file
+    available = set(io_gwf.iter_channel_names(gwfcache[0]))
+    if len(gwfcache) > 1:
+        available.update(io_gwf.iter_channel_names(gwfcache[-1]))
```
Good catch! This is implemented in the latest commit.

alurban commented:
@duncanmmacleod, thanks for posting comments, I've addressed each of your requested changes (all of which I agreed with).

duncanmmacleod commented on bin/gwdetchar-scattering:
```diff
@@ -291,7 +291,10 @@ for i, channel in enumerate(sorted(allchannels)):
+        try:
+            ts = alldata[j][channel]
+        except KeyError:
+            warnings.warn("Skipping {}".format(channel), UserWarning)
```
I think you need a `continue` here.

alurban replied:
This is done in the last commit.

duncanmmacleod commented on gwdetchar/saturation.py:
```diff
@@ -187,6 +187,10 @@ def is_saturated(channel, cache, start=None, end=None, indicator='LIMEN',
+
+    # check limits for returned channels
+    channels = [
+        c for c in channels if '{}_{}'.format(c, indicator) in data.keys()]
```
```suggestion
        c for c in channels if '{}_{}'.format(c, indicator) in data]
```

alurban commented:
@duncanmmacleod, I've addressed your review comments, and assembled test output of affected tools (all require `LIGO.ORG` credentials):

Tool | Date Analyzed | Output | Log | Notes
 -- | -- | -- | -- | --
`gwdetchar-scattering` | 2017-08-17 | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/scattering/day/20170817/) | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/scattering/day/20170817/scattering.out) | Day during O2, SQZ suspension optic channels didn't exist yet
`gwdetchar-overflow` | 2017-08-17 | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/overflows/day/20170817/) | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/overflows/day/20170817/overflows.out) | Day during O2
`gwdetchar-omega` | 2019-01-08 | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/hveto/day/20190108/scans/1230955301.876464/) | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/hveto/day/20190108/scans/logs/gwdetchar-omega_1230955301_876464.error) | Various channels didn't exist in January, output column links to `stderr` output
`gwdetchar-software-saturations` | 2019-03-12 | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/software-saturations/day/20190312/) | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/software-saturations/day/20190312/saturations.out) | Run without the `--skip` command line flag, hence several nonexistent channels
`gwdetchar-lasso-correlation` | 2019-03-13 | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/lasso/full-results/day/20190313/) | [**Link**](https://ldas-jobs.ligo-la.caltech.edu/~aurban/lasso/full-results/day/20190313/lasso.out) |
